### PR TITLE
`Development`: rewrite FileService methods to use Path instead of String for its parameters

### PR DESCRIPTION
**Pull request by 1999Markus**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

https://github.com/ls1intum/Artemis/issues/5793

### Description

I changed the types of the parameters of the function `renameDirectory` in `src/main/java/de/tum/in/www1/artemis/service/FileService.java` from `String` to `Path`. I also added the test `testRenameDirectory()` in `de/tum/in/www1/artemis/service/FileServiceTest.java` to test the function behaves as expected.

### Steps for Testing

Prerequisites:

1. Log in to Artemis as an Instructor
2. Create a Programming Exercise

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### CODE REVIEW

**Comment by 1999Markus** [comment]
How do I test this on a test server?

File: src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java [hunk 1]
```diff
@@ -311,4 +311,17 @@ void testDeleteFiles_shouldNotThrowException() {
...
         mockedFiles.close();
     }
+
+    @Test
+    void testRenameDirectory() {
+        // File is deleted automatically after test
+        writeFile("some-path/some-file", "some-content");
+        Path source_dir = Path.of("./exportTest/some-path");
+        Path target_dir = Path.of("./exportTest/some-path2");
+        Path target_file = Path.of("./exportTest/some-path2/some-file");
```
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java:321]:
```suggestion
        Path sourceDir = Path.of("./exportTest/some-path");
        Path targetDir = Path.of("./exportTest/some-path2");
        Path targetFile = Path.of("./exportTest/some-path2/some-file");
```

**Review by Strohgelaender — CHANGES_REQUESTED** [review]
Thanks for the changes.

While this improves the APi of the FileService, by just calling `toFile` we still use the old API inside of the method. This should also be changed to use the `nio` methods.

File: src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java (hunk 1 above)
Comment (simon-hng) [reply src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java:321]: Yes I agree, there are also still usages of strings in the public API.
We should duplicate the method header, and mark it as `@deprecated`, then use delegation to delegate to the Method with the signature we want to be used. i.e.

```java
/**
 * @deprecated. use method ...
 */
public void methodToBeRefactored(String path) {
  methodToBeRefactored(Path.from(path));
}

public void methodToBeRefactored(Path path) {
  // Logic moved here.
}

```

Comment (simon-hng) [reply src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java:321]: In case it wasn't clear. This was just a suggestion how I would phase out deprecated code

Comment (1999Markus) [reply src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java:321]: So I should focus merely on the API of the FileService and leave the responsibility for calling the correct method to the caller ( e.g another service ) ?

Comment (1999Markus) [reply src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java:321]: > While this improves the APi of the FileService, by just calling `toFile` we still use the old API inside of the method. This should also be changed to use the `nio` methods.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 2]
```diff
@@ -487,17 +490,18 @@ public void copyResources(Resource[] resources, String prefix, String targetDire
...
                 continue;
             }
-
-            Path copyPath = Path.of(targetDirectoryPath + targetFilePath);
-            File parentFolder = copyPath.toFile().getParentFile();
-            if (!parentFolder.exists()) {
-                Files.createDirectories(parentFolder.toPath());
+            // will append the targetFilePath to the targetDirectoryPath
+            Path copyPath = targetDirectoryPath.resolve(targetFilePath);
+            // File parentFolder = copyPath.toFile().getParentFile();
```
Comment (simon-hng) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java:495]: You should remove commented out code. -> Lava flow antipattern

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 3]
```diff
@@ -474,30 +475,33 @@ private File generateTargetFile(String originalFilename, String targetFolder, Bo
...
+            Path parentFolder = copyPath.getParent();
+            if (!Files.exists(parentFolder)) {
+                Files.createDirectories(parentFolder);
             }
 
             Files.copy(resource.getInputStream(), copyPath, REPLACE_EXISTING);
             // make gradlew executable
             if (targetFilePath.endsWith("gradlew")) {
-                copyPath.toFile().setExecutable(true);
+                Files.setPosixFilePermissions(copyPath, Set.of(OWNER_EXECUTE, OWNER_READ, OWNER_WRITE));
```
Comment (simon-hng) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java:504]: Are read/write also set in the previous implementation?

Comment (1999Markus) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:504]: Not 100% sure, I'll rewrite this without assuming anything.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 4]
```diff
@@ -541,14 +546,15 @@ private boolean isIgnoredDirectory(final String filePath) {
      * @param targetDirectoryPath the path of the folder where the renamed folder should be located
      * @throws IOException if the directory could not be renamed.
      */
-    public void renameDirectory(String oldDirectoryPath, String targetDirectoryPath) throws IOException {
-        File oldDirectory = new File(oldDirectoryPath);
+
+    public void renameDirectory(Path oldDirectoryPath, Path targetDirectoryPath) throws IOException {
+        File oldDirectory = oldDirectoryPath.toFile();
```
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java:551]: why do still use toFile here? We should use Files.exists instead in the if-condition

File: src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java [hunk 5]
```diff
@@ -298,4 +298,17 @@ void testDeleteFiles_shouldNotThrowException() {
         Path path = Path.of("some-random-path-which-does-not-exist");
         assertDoesNotThrow(() -> fileService.deleteFiles(List.of(path)));
     }
+
+    @Test
+    void testRenameDirectory() {
+        // File is deleted automatically after test
```
Comment (tobias-lippert) [inline src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java:304]: the test should also not use toFile.
and where is the file deleted again?
we should probably also assert that the the path of source dir equals the one of target dir after the rename.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 6]
```diff
@@ -474,30 +476,33 @@ private File generateTargetFile(String originalFilename, String targetFolder, Bo
...
+            Path parentFolder = copyPath.getParent();
+            if (!Files.exists(parentFolder)) {
+                Files.createDirectories(parentFolder);
             }
 
             Files.copy(resource.getInputStream(), copyPath, REPLACE_EXISTING);
             // make gradlew executable
             if (targetFilePath.endsWith("gradlew")) {
-                copyPath.toFile().setExecutable(true);
+                Files.getPosixFilePermissions(copyPath).add(PosixFilePermission.OWNER_EXECUTE);
```
Comment (ge65cer) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java:505]: Please [take a look](https://github.com/ls1intum/Artemis/pull/5992/files) at the Codacy warning here.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 7]
```diff
@@ -474,30 +476,33 @@ private File generateTargetFile(String originalFilename, String targetFolder, Bo
...
+
         for (Resource resource : resources) {
             // Replace windows separator with "/"
             String fileUrl = java.net.URLDecoder.decode(resource.getURL().toString(), StandardCharsets.UTF_8).replaceAll("\\\\", "/");
             // cut the prefix (e.g. 'exercise', 'solution', 'test') from the actual path
             int index = fileUrl.indexOf(prefix);
 
-            String targetFilePath = keepParentFolder ? fileUrl.substring(index + prefix.length()) : "/" + resource.getFilename();
+            // We use prefix.length() + 1 to also cut the "/" after the prefix
+            String targetFilePath = keepParentFolder ? fileUrl.substring(index + prefix.length() + 1) : resource.getFilename();
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java:489]: Can the `prefix` be empty and the `fileUrl` start without `/`?

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 8]
```diff
@@ -640,7 +641,7 @@
         if (startPath.contains(targetString)) {
             log.debug("Target String found, replacing..");
             String targetPath = startPath.replace(targetString, replacementString);
-            renameDirectory(startPath, targetPath);
+            renameDirectory(Path.of(startPath), Path.of(targetPath));
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:649]: Code scanning alert

**Review by julian-christl — CHANGES_REQUESTED** [review]

**Review by b-fein — COMMENTED** [review]
Thanks for tackling a code quality issue!

I think there was some confusion about using `java.io` vs. `java.nio` methods: The main intention of the linked issue is to replace cases where paths are represented as `String`s with manually appending `/` and other path parts with the proper `Path`/`File` types.
If then the regular `java.io` methods are useful/short/easy to understand for those, then it is perfectly fine to keep using them.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (1999Markus) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:649]: Will use `Paths.get()` which sanitizes the user input

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java (hunk 7 above)
Comment (1999Markus) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:489]: It is currently not forbidden. This could lead to cutting the first letter from a path.

File: src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java (hunk 5 above)
Comment (1999Markus) [reply src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java:304]: The test class features a @AfterEach function which deletes all files `./exportTest`. Which is where the generated files are located. So everything is taken care of.

Comment (1999Markus) [reply src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java:304]: Why should the path of the sourceDir equal the targetDir?

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 9]
```diff
@@ -640,7 +645,7 @@
         if (startPath.contains(targetString)) {
             log.debug("Target String found, replacing..");
             String targetPath = startPath.replace(targetString, replacementString);
-            renameDirectory(startPath, targetPath);
+            renameDirectory(Paths.get(startPath).normalize(), Paths.get(targetPath).normalize());
```
[the remark this answers is not in the document]
Comment (1999Markus) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:648]: This is as sanitized as possible using` java.nio`.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 10]
```diff
@@ -474,24 +475,27 @@ private File generateTargetFile(String originalFilename, String targetFolder, Bo
      * @param keepParentFolder    if true also creates the resources with the folder they are currently in (e.g. current/parent/* -> new/parent/*)
      * @throws IOException if the copying operation fails.
      */
-    public void copyResources(Resource[] resources, String prefix, String targetDirectoryPath, Boolean keepParentFolder) throws IOException {
+    public void copyResources(Resource[] resources, String prefix, Path targetDirectoryPath, boolean keepParentFolder) throws IOException {
         for (Resource resource : resources) {
             // Replace windows separator with "/"
-            String fileUrl = java.net.URLDecoder.decode(resource.getURL().toString(), StandardCharsets.UTF_8).replaceAll("\\\\", "/");
+            String fileUrl = java.net.URLDecoder.decode(resource.getURL().toString(), StandardCharsets.UTF_8).replace("\\\\", "/");
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java:481]: Why don't we import the `URLDecoder`?
And why the change from replaceAll to replace?

Comment (1999Markus) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:481]: Using `replace` was suggested by @b-fein since we are not using a regex. Regarding the `URLDecoder` you're right one can simply import it and make the code less verbose

File: src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java (hunk 5 above)
Comment (tobias-lippert) [reply src/test/java/de/tum/in/www1/artemis/service/FileServiceTest.java:304]: that's what I would expect from the name. It renames the sourceDir to targetDir.
